cps/kobo.py

```diff
@@ -392,6 +394,222 @@ def get_metadata(book):
+        abort(400, description="Malformed tags POST request. Data is missing 'Items' field")
+
+    shelf = ub.session.query(ub.Shelf).filter(and_(ub.Shelf.uuid) == tag_id,
+                                              ub.Shelf.user_id == current_user.id).one_or_none()
+    if not shelf:
+        log.debug("Received Kobo request on a collection unknown to CalibreWeb")
+        abort(404, description="Collection isn't known to CalibreWeb")
+
+    if not shelf_lib.check_shelf_edit_permissions(shelf):
+        abort(401, description="User is unauthaurized to edit shelf.")
+
+    items_unknown_to_calibre = add_items_to_shelf(items, shelf)
+    if items_unknown_to_calibre:
+        log.debug("Received request to add an unknown book to a collecition. Silently ignoring item.")
+
+    ub.session.merge(shelf)
+    ub.session.commit()
+
+    return make_response('', 201)
+
+
+@kobo.route("/v1/library/tags/<tag_id>/items/delete", methods=["POST"])
+@login_required
+def HandleTagRemoveItem(tag_id):
+    tag_request = request.json
+    items = None
+    try:
+        items = tag_request["Items"]
+    except KeyError:
+        log.debug("Received malformed v1/library/tags/<tag_id>/items/delete request.")
+        abort(400, description="Malformed tags POST request. Data is missing 'Items' field")
+
+    shelf = ub.session.query(ub.Shelf).filter(ub.Shelf.uuid == tag_id,
+                                              ub.Shelf.user_id == current_user.id).one_or_none()
+    if not shelf:
+        log.debug(
+            "Received a request to remove an item from a Collection unknown to CalibreWeb.")
+        abort(404, description="Collection isn't known to CalibreWeb")
+
+    if not shelf_lib.check_shelf_edit_permissions(shelf):
+        abort(401, description="User is unauthaurized to edit shelf.")
+
+    items_unknown_to_calibre = []
+    for item in items:
+        if item["Type"] != "ProductRevisionTagItem":
+            items_unknown_to_calibre.append(item)
+            continue
+
+        book = db.session.query(db.Books).filter(db.Books.uuid == item["RevisionId"]).one_or_none()
+        if not book:
+            items_unknown_to_calibre.append(item)
+            continue
+
+        shelf.books.filter(ub.BookShelf.book_id == book.id).delete()
+    ub.session.commit()
+
+    if items_unknown_to_calibre:
+        log.debug("Received request to remove an unknown book to a collecition. Silently ignoring item.")
+
+    return make_response('', 200)
+
+
+# Add new, changed, or deleted shelves to the sync_results.
+# Note: Public shelves that aren't owned by the user aren't supported.
+def sync_shelves(sync_token, sync_results):
+    new_tags_last_modified = sync_token.tags_last_modified
+
+    for shelf in ub.session.query(ub.ShelfArchive).filter(func.datetime(ub.ShelfArchive.last_modified) > sync_token.tags_last_modified, ub.ShelfArchive.user_id == current_user.id):
+        new_tags_last_modified = max(shelf.last_modified, new_tags_last_modified)
```

Comment:
Haven't fully understood what we are doing there.
I did the following: Generate shelf on UI (private one), sync it to kobo (New Tag is synced), add book to shelf, try to sync it again.
Expected result: Changed tag is synced, but I get empty response.
Upon looking into ShelfArchive table I find this:
![grafik](https://user-images.githubusercontent.com/6438895/79894317-90f68780-8405-11ea-9ef9-8a55bad141cb.png)
User ID is identical to the uuid  but it's compared to the internal Calibre-Web user iD, which is a number from 1 to x, so comparsion fails, and the book in shelf isn't synced.

Reply:
This particular section is propagating deletions of shelves from CalibreWeb to the device. And indeed it looks like there's a bug in the delete_shelf_helper() I added to shelf.py; specifically line 286
```
    ub.session.add(ub.ShelfArchive(uuid = cur_shelf.uuid, user_id = cur_shelf.uuid))
```
With this bug, deleting a shelf from CalibreWeb and syncing the device doesn't remove the shelf form the device.
With 9296d35 , deleting a shelf from CalibreWeb will create an entry in the ShelfArchive table with the correct user_id field. Deleting a shelf form CalibreWeb also removes it from the device after syncing.